Get following error in VS: error MSB4018: The “ResolvePackageAssets” task failed unexpectedly.

```
[severity:Other]
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: The “ResolvePackageAssets” task failed unexpectedly.
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: System.ArgumentException: An item with the same key has already been added.
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at System.Collections.Generic.Dictionary2.Insert(TKey key, TValue value, Boolean add) 61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018:    at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable1 source, Func2 keySelector, Func2 elementSelector, IEqualityComparer`1 comparer)
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheWriter.ComputePackageExclusions()
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheWriter…ctor(ResolvePackageAssets task)
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader.CreateReaderFromDisk(ResolvePackageAssets task, Byte[] settingsHash)
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader…ctor(ResolvePackageAssets task)
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ReadItemGroups()
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ExecuteCore()
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.NET.Build.Tasks.TaskBase.Execute()
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
61>C:\Program Files\dotnet\sdk\3.1.400\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error MSB4018: at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
61> 0 Warning(s)
61> 1 Error(s)
```

--- a/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolvePackageAssetsTask.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolvePackageAssetsTask.cs
@@ -6,9 +6,11 @@ using Microsoft.Build.Framework;
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using System.IO;
 using System.Linq;
 using System.Reflection;
 using Xunit;
+using static Microsoft.NET.Build.Tasks.ResolvePackageAssets;
 
 namespace Microsoft.NET.Build.Tasks.UnitTests
 {
@@ -80,6 +82,39 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
 
             newHash.Should().BeEquivalentTo(oldHash,
                 because: $"{nameof(task.DesignTimeBuild)} should not be included in hash.");
+        }
+
+        [Fact]
+        public void It_does_not_error_on_duplicate_package_names()
+        {
+            string projectAssetsJsonPath = Path.GetTempFileName();
+            var assetsContent =  @"{
+  `version`: 3,
+  `targets`: {
+    `net5.0`: {
+      `Humanizer.Core/2.2.0`: {
+        `type`: `package`
+      },
+      `Humanizer.Core/2.2.1`: {
+        `type`: `package`
+      }
+    }
+  },
+  `project`: {
+    `version`: `1.0.0`,
+    `frameworks`: {
+      `net5.0`: {
+        `targetAlias`: `net5.0`
+      }
+    }
+  }
+}".Replace('`', '"');
+            File.WriteAllText(projectAssetsJsonPath, assetsContent);
+
+            var task = InitializeTask(out _);
+            task.ProjectAssetsFile = projectAssetsJsonPath;
+            task.TargetFramework = "net5.0";
+            new CacheWriter(task); // Should not error
         }
 
         private ResolvePackageAssets InitializeTask(out IEnumerable<PropertyInfo> inputProperties)

--- a/src/Tasks/Microsoft.NET.Build.Tasks/LockFileCache.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/LockFileCache.cs
@@ -36,12 +36,12 @@ namespace Microsoft.NET.Build.Tasks
             string lockFileKey = GetTaskObjectKey(path);
 
             LockFile result;
-            object existingLockFileTaskObject = _buildEngine.GetRegisteredTaskObject(lockFileKey, RegisteredTaskObjectLifetime.Build);
+            object existingLockFileTaskObject = _buildEngine?.GetRegisteredTaskObject(lockFileKey, RegisteredTaskObjectLifetime.Build);
             if (existingLockFileTaskObject == null)
             {
                 result = LoadLockFile(path);
 
-                _buildEngine.RegisterTaskObject(lockFileKey, result, RegisteredTaskObjectLifetime.Build, true);
+                _buildEngine?.RegisterTaskObject(lockFileKey, result, RegisteredTaskObjectLifetime.Build, true);
             }
             else
             {

--- a/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
+++ b/src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs
@@ -618,7 +618,7 @@ namespace Microsoft.NET.Build.Tasks
             }
         }
 
-        private sealed class CacheWriter : IDisposable
+        internal sealed class CacheWriter : IDisposable
         {
             private const int InitialStringTableCapacity = 32;
 
@@ -1347,7 +1347,10 @@ namespace Microsoft.NET.Build.Tasks
             private void ComputePackageExclusions()
             {
                 var copyLocalPackageExclusions = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
-                var libraryLookup = _runtimeTarget.Libraries.ToDictionary(p => p.Name, StringComparer.OrdinalIgnoreCase);
+                var libraryLookup = _runtimeTarget.Libraries
+                    .GroupBy(p => p.Name)
+                    .Select(g => g.First())
+                    .ToDictionary(p => p.Name, StringComparer.OrdinalIgnoreCase);
 
                 // Only exclude platform packages for framework-dependent applications
                 if ((!_task.IsSelfContained || string.IsNullOrEmpty(_runtimeTarget.RuntimeIdentifier)) &&

